File: src/main/java/no/novari/kafka/topic/configuration/EntityTopicConfigurationMappingService.java
package no.novari.kafka.topic.configuration;

import org.springframework.stereotype.Service;

@Service
public class EntityTopicConfigurationMappingService {

    public TopicConfiguration toTopicConfiguration(EntityTopicConfiguration entityTopicConfiguration) {
        TopicConfiguration.TopicConfigurationBuilder topicConfigurationBuilder = TopicConfiguration.builder();

        topicConfigurationBuilder.partitions(entityTopicConfiguration.getPartitions());

        entityTopicConfiguration.getLastValueRetentionTime()
                .ifPresent(retentionTime ->
                        topicConfigurationBuilder.deleteCleanupPolicy(
                                TopicDeleteCleanupPolicyConfiguration
                                        .builder()
                                        .retentionTime(retentionTime)
                                        .build()
                        )
                );

        TopicCompactCleanupPolicyConfiguration.TopicCompactCleanupPolicyConfigurationBuilder
                compactCleanupPolicyTopicConfigurationBuilder = TopicCompactCleanupPolicyConfiguration.builder();

        compactCleanupPolicyTopicConfigurationBuilder
                .maxCompactionLag(entityTopicConfiguration.getCleanupFrequency().getMaxCompactionLag());

        compactCleanupPolicyTopicConfigurationBuilder.nullValueRetentionTime(
                entityTopicConfiguration.getNullValueRetentionTime()
        );
        topicConfigurationBuilder.compactCleanupPolicy(compactCleanupPolicyTopicConfigurationBuilder.build());


        TopicSegmentConfiguration.TopicSegmentConfigurationBuilder segmentTopicConfiguration =
                TopicSegmentConfiguration.builder();

        segmentTopicConfiguration.openSegmentDuration(
                entityTopicConfiguration.getCleanupFrequency().getSegmentDuration()
        );
        topicConfigurationBuilder.segmentConfiguration(segmentTopicConfiguration.build());


        return topicConfigurationBuilder.build();
    }
}

File: src/main/kotlin/no/novari/kafka/topic/configuration/EntityTopicConfigurationMappingService.kt
package no.novari.kafka.topic.configuration

import org.springframework.stereotype.Service

@Service
class EntityTopicConfigurationMappingService {
    fun toTopicConfiguration(entityTopicConfiguration: EntityTopicConfiguration): TopicConfiguration {
        val topicConfigurationBuilder = TopicConfiguration.builder()
        topicConfigurationBuilder.partitions(entityTopicConfiguration.partitions)

        entityTopicConfiguration.lastValueRetentionTime.ifPresent { retentionTime ->
            topicConfigurationBuilder.deleteCleanupPolicy(
                TopicDeleteCleanupPolicyConfiguration.builder().retentionTime(retentionTime).build(),
            )
        }

        topicConfigurationBuilder.compactCleanupPolicy(
            TopicCompactCleanupPolicyConfiguration
                .builder()
                .maxCompactionLag(entityTopicConfiguration.cleanupFrequency.maxCompactionLag)
                .nullValueRetentionTime(entityTopicConfiguration.nullValueRetentionTime)
                .build(),
        )

        topicConfigurationBuilder.segmentConfiguration(
            TopicSegmentConfiguration
                .builder()
                .openSegmentDuration(entityTopicConfiguration.cleanupFrequency.segmentDuration)
                .build(),
        )

        return topicConfigurationBuilder.build()
    }
}
